### DISCUSSION TITLE
Path parameter (enum) with class and nested dependency resolution

### DISCUSSION THREAD

**Discussion by nb-programmer** [fastapi/fastapi#9793] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Depends
from enum import Enum

# Somewhere here will import get_db_session for SQLAlchemy DB which is a generator dependency (yield)

app = FastAPI()

class ServiceHandlerBase:
    def work(self): return ''

class Service1Handler(ServiceHandlerBase):
    def __init__(self, db = Depends(get_db_session)):
        self.session = db
    def work(self) -> str:
        self.session.commit() # Fails
        return 'service 1 did work! I Got DB as well!'


class Service2Handler(ServiceHandlerBase):
    def __init__(self):
        pass
    def work(self) -> str:
        return "service 2 did work! I don't need a db!"


class MyServiceType(str, Enum):
    service1 = Service1Handler
    service2 = Service2Handler

@app.get("/{service:str}")
async def my_service_selector(service: MyServiceType) -> str:
    print(service, type(service))
    return await service.value().work() # Dependency of service1 does not resolve

# * run app like normal using uvicorn*
```

### Description

I want to add a dependency that can be selected based on the path parameter (`service` in this case). Like normal how you add a dependency as parameter using `abc: type = Depends(ClassName)`, but the ClassName must be resolved by the value of service, so that `abc` is an instance of either `Service1Handler` or `Service2Handler` from user input.

Why I want to do this: Obviously I could just add 2 routes with /service1 and /service2 to separately handle each case. But I was thinking if I could generalize this using Path Parameters.
The Enum code I posted above works but doesn't resolve sub-dependencies. I don't know any other way in FastAPI to do it the way I want.

I want to do it this way:
```python
def get_service_class(name: MyServiceType ) -> Type[ServiceHandlerBase]:
    return Depends(name.value)

@app.route('/{service:str}')
async def service_select(service: ServiceHandlerBase = Depends(get_service_class(service))):
    # work...

```

So that a dropdown is shown in SwaggerUI

### Operating System

Windows

### FastAPI Version

0.95.1

### Python Version

3.9.13

**Comment by ahmedabdou14** [comment, marked as answer]
You can easily do it using an external handler method that will read the path parameter and return the correct service.

```python
from typing import Annotated
from fastapi import Depends, APIRouter, Path
from enum import Enum

router = APIRouter()


class ServiceHandlerBase:
    def work(self):
        return ""


class Service1Handler(ServiceHandlerBase):
    def __init__(self):
        pass

    def work(self) -> str:
        return "service 1 did work! I have a db!"


class Service2Handler(ServiceHandlerBase):
    def __init__(self):
        pass

    def work(self) -> str:
        return "service 2 did work! I don't need a db!"


class MyServiceType(str, Enum):
    service1 = "service1"
    service2 = "service2"


def get_service(
    service: Annotated[MyServiceType, Path(...)]
) -> ServiceHandlerBase:
    if service == MyServiceType.service1:
        return Service1Handler()
    elif service == MyServiceType.service2:
        return Service2Handler()
    else:
        raise Exception("Unknown service")


ServiceDep = Annotated[ServiceHandlerBase, Depends(get_service)]


@router.get("/{service}")
async def my_service_selector(service: ServiceDep) -> str:
    print(service, type(service))
    return service.work()
```
